internals/tpl/template_test.go

```diff
@@ -47,105 +46,116 @@ func TestParse(t *testing.T) {
 		},
 		"two": {
 			raw:      fmt.Sprintf(`${ %s }${ %s}`, testSecretPath, testSecretPath2),
-			expected: []string{testSecretPath, testSecretPath2},
+			expected: []node{key(testSecretPath), key(testSecretPath2)},
 		},
 		"duplicates": {
 			raw:      fmt.Sprintf(`${ %s }${ %s}${%s }`, testSecretPath, testSecretPath2, testSecretPath2),
-			expected: []string{testSecretPath, testSecretPath2},
+			expected: []node{key(testSecretPath), key(testSecretPath2), key(testSecretPath2)},
 		},
 		"invalid_path": {
 			raw:      `${ invalidpath }`,
-			expected: []string{"invalidpath"},
+			expected: []node{key("invalidpath")},
 		},
 		"empty": {
 			raw:      `${}`,
-			expected: []string{""},
+			expected: []node{key("")},
 		},
 		"empty_nested": {
 			raw:      `${${}}`,
-			expected: []string{"${"},
+			expected: []node{key("${"), val("}")},
 		},
 		"path_folowed_by_delim": {
 			raw:      fmt.Sprintf(`${ %s ${}}`, testSecretPath),
-			expected: []string{fmt.Sprintf("%s ${", testSecretPath)},
+			expected: []node{key(fmt.Sprintf("%s ${", testSecretPath)), val("}")},
 		},
 		"path_followed_by_nested_path": {
 			raw:      fmt.Sprintf(`${ %s ${ %s }}`, testSecretPath, testSecretPath2),
-			expected: []string{fmt.Sprintf("%s ${ %s", testSecretPath, testSecretPath2)},
+			expected: []node{key(fmt.Sprintf("%s ${ %s", testSecretPath, testSecretPath2)), val("}")},
 		},
 		"nested": {
 			raw:      fmt.Sprintf(`${ ${ %s }}`, testSecretPath),
-			expected: []string{fmt.Sprintf("${ %s", testSecretPath)},
+			expected: []node{key(fmt.Sprintf("${ %s", testSecretPath)), val("}")},
 		},
 		"unclosed": {
 			raw: `${ foobar`,
 			err: ErrTagNotClosed(DefaultEndDelimiter),
 		},
 		"unopened": {
 			raw:      `{ foobar }`,
-			expected: []string{},
+			expected: []node{val("{ foobar }")},
 		},
 		"unclosed_with_nested": {
 			raw:      fmt.Sprintf(`${ ${ %s }`, testSecretPath),
-			expected: []string{fmt.Sprintf("${ %s", testSecretPath)},
+			expected: []node{key(fmt.Sprintf("${ %s", testSecretPath))},
 		},
 		"unclosed_with_empty_nested": {
 			raw:      `${ ${}`,
-			expected: []string{"${"},
+			expected: []node{key("${")},
 		},
 		"unclosed_with_path_and_empty_nested": {
 			raw:      fmt.Sprintf(`${ %s ${}`, testSecretPath),
-			expected: []string{fmt.Sprintf("%s ${", testSecretPath)},
+			expected: []node{key(fmt.Sprintf("%s ${", testSecretPath))},
 		},
 		"unclosed_with_path_and_nested": {
 			raw:      fmt.Sprintf(`${ %s ${ %s }`, testSecretPath, testSecretPath2),
-			expected: []string{fmt.Sprintf("%s ${ %s", testSecretPath, testSecretPath2)},
+			expected: []node{key(fmt.Sprintf("%s ${ %s", testSecretPath, testSecretPath2))},
 		},
 		"YAML": {
-			raw:      dataYAML,
-			expected: []string{testSecretPath},
+			raw: dataYAML,
+			expected: []node{
+				val(`config:
+			some_field: "some value"
+			secret_field: "`),
+				key(testSecretPath),
+				val("\"")},
 		},
 		"JSON": { // TODO: Decide what to do in this case
```

Comment:
What's this?

Reply:
Ah, we can remove that comment.

This was about whether we'd want to check for closing delimiters that are not opened, which occur in JSON a lot when you have '${` and `}` as delimiters.

We decided not to check for closing delimiters when no opening delimiter occurred. Just like Go, Liquid and CUE templates do.